### PR TITLE
The second NonlinearSystem hierarchy

### PR DESCRIPTION
**Pull request by fdkong**
Our goal is:

```
SystemBase -> SystemTemp<> +--> AuxiliarySystem
                           |
                           +--> NonlinearSystem  <---+ AssemblySystem 
                           |                         |                
                           +--> MooseEigenSystem <----
```

What I did:

(1)  Renamed `NonlinearSystem` to `AssemblySystem`,  and created  en empty `NonlinearSystem` class.
     `AssemblySystem` will work as a base class of `NonlinearSystem` and `NonlinearEigenSystem`.

(2)  Moved `SystemTempl` into `NonlinearSystem`  to make `AssemblySystem` independent.
      Any functions which depends on `SysBase` and `SystemTempl`, are moved back to `NonlinearSystem`.

Conclusions:
Most of functions have to be moved back to `NonlinearSystem` again.  2187 lines of the code have been moved back to `NonlinearSystem`, and only 539 lines of the code are left in `AssemblySystem`.

It is, possibly, not a good idea to design in this way, (I honestly don't know). 

I am starting to think about the first proposal:

```
SystemBase -> SystemTemp<> +--> AuxiliarySystem
                           |
                           +-->IntermediateBase+--> NonlinearSystem
                                               |
                                               +--> MooseEigenSystem
```

   This PR is NOT ready to merge, and I am here just trying to show you guys what issues are.

### CODE REVIEW

File: framework/include/base/NonlinearSystem.h [hunk 1]
```diff
@@ -15,76 +15,72 @@
...
+  virtual void serializeSolution();
+  virtual NumericVector<Number> & serializedSolution() override;
+
+  virtual NumericVector<Number> & residualCopy() override;
+  virtual NumericVector<Number> & residualGhosted() override;
+
+  virtual void augmentSparsity(SparsityPattern::Graph & sparsity,
+                               std::vector<dof_id_type> & n_nz,
+                               std::vector<dof_id_type> & n_oz) override;
+
```
Comment (fdkong) [inline framework/include/base/NonlinearSystem.h]: These functions override the ones from  `SysBase` and `SystemTempl`.

File: framework/src/base/NonlinearSystem.C [hunk 2]
```diff
@@ -294,6 +222,153 @@ NonlinearSystem::restoreSolutions()
...
+
+    const dof_id_type first_dof_on_proc = dofMap().first_dof(processor_id());
+    const dof_id_type end_dof_on_proc   = dofMap().end_dof(processor_id());
+
+    // If we're on the last processor then we have all of the scalar dofs
+    const dof_id_type n_scalar_dofs_on_proc = processor_id() == n_processors()-1 ? dofMap().n_SCALAR_dofs() : 0;
+
+    // If we're _not_ on the last processor then all of the scalar dofs are off processor (n_oz)
+    const dof_id_type n_scalar_dofs_not_on_proc = processor_id() == n_processors()-1 ? 0 : dofMap().n_SCALAR_dofs();
+
```
Comment (fdkong) [inline framework/src/base/NonlinearSystem.C]: processor_id() from libmesh::parallelobject

File: framework/src/base/NonlinearSystem.C [hunk 3]
```diff
@@ -294,6 +222,153 @@ NonlinearSystem::restoreSolutions()
...
+    // The total number of dofs on and off processor (including scalar dofs)
+    const dof_id_type n_dofs_on_proc = dofMap().n_local_dofs() + n_scalar_dofs_on_proc;
+    const dof_id_type n_dofs_not_on_proc = (dofMap().n_dofs() - dofMap().n_local_dofs()) + n_scalar_dofs_not_on_proc;
+
```
Comment (fdkong) [inline framework/src/base/NonlinearSystem.C]: dofMap() from `SysBase`

File: framework/src/base/NonlinearSystem.C [hunk 4]
```diff
@@ -625,100 +581,27 @@ NonlinearSystem::addBoundaryCondition(const std::string & bc_name, const std::st
...
+  for (THREAD_ID tid = 0; tid < libMesh::n_threads(); ++tid)
   {
-    MooseSharedPointer<Damper> damper = _factory.create<Damper>(damper_name, name, parameters, tid);
+    MooseSharedPointer<InterfaceKernel> interface_kernel = _factory.create<InterfaceKernel>(interface_kernel_name, name, parameters, tid);
 
-    // Attempt to cast to the damper types
-    MooseSharedPointer<ElementDamper> ed = MooseSharedNamespace::dynamic_pointer_cast<ElementDamper>(damper);
-    MooseSharedPointer<GeneralDamper> gd = MooseSharedNamespace::dynamic_pointer_cast<GeneralDamper>(damper);
+    const std::set<BoundaryID> & boundary_ids = interface_kernel->boundaryIDs();
+    _vars[tid].addBoundaryVar(boundary_ids, &interface_kernel->variable());
```
Comment (fdkong) [inline framework/src/base/NonlinearSystem.C]: `_vars` from `SystemBase`

**Comment by fdkong** [comment]
All functions in `NonlinearSystem` depend on the `SysBase` and `SystemTempl`, and it is not easy to raise them to `AssemblySystem` .

**Comment by fdkong** [comment]
@andrsd, How do you think? Any suggestions to resolve this?  Might also try the first proposal too ?

Thanks a lot.

**Comment by andrsd** [comment]
I will be able to get to this sometime tomorrow...

**Comment by YaqiWang** [comment]
i am guessing we need to write moose2 now.

**Comment by permcody** [comment]
Moose3! David's design of the assembly and lower level objects is the
second iteration.

**Comment by fdkong** [comment]
@andrsd, there are three possibilities: (1) Instead of inheriting from `AssemblySystem`, we could just use `AssemblySystem` directly in `NonlinearSystem` and `NonlinearEigenSystem`; (2) Another possible solution is to change the `SystemTmpl` to a templete-independent class. We could call ` subproblem.es().add_system` in the children classes and then pass the corresponding "sys" back to `SystemTmpl`.

The third one is to develop MOOSE-3?

**Comment by andrsd** [comment]
I do not think we need to start working on MOOSE 3.0. I am working on this and putting SLEPc in. It does take a more involved way, but I think it will work... When I have something to share, I will do so... Meanwhile, stay tuned...

**Comment by fdkong** [comment]
> I do not think we need to start working on MOOSE 3.0. I am working on this and putting SLEPc in. It does take a more involved way, but I think it will work... When I have something to share, I will do so... Meanwhile, stay tuned...

Thanks. You call SLEPc directly or all ways through SLEPc?

**Comment by andrsd** [comment]
I will use it via libMesh's interface

**Pull request closed without merging**